### PR TITLE
Small Rework for Frontier Distance Calculation

### PR DESCRIPTION
**Pull request by Shawn8901**
As the Current implementation is able to override a frontier distance (mid/near), this is should include a fix for that.
As i had a save game with quite a lot military buildings, i recognized some kind of lag after changing state of a military builiding (p.e. if it gets destroyed). 
After looking at the code, i realized that the path-calculation is done for every military building, which is quite useless. it should only override the state if its mid/near. 
So i moved to code.

Also i have added some basic refactoring, so i am setting the distance on the other building at the end of the loop, instead of in every if-clause, since its redundant code.

### CODE REVIEW

File: src/buildings/nobMilitary.cpp [hunk 1]
```diff
@@ -329,60 +329,49 @@ void nobMilitary::LookForEnemyBuildings(const nobBaseMilitary* const exception)
...
+            {
+                frontier_distance = newFrontierDistance;
+            }
+
+            // set calculated frontier distance to checked building.
+            if(BuildingProperties::IsMilitary((*it)->GetBuildingType()))
+                static_cast<nobMilitary*>(*it)->NewEnemyMilitaryBuilding(newFrontierDistance);
         }
+    }
+    // check for habor points
```
Comment (Flow86) [inline src/buildings/nobMilitary.cpp]: typo "habor" ;-)

**Comment by Flamefire** [comment]
The changes look reasonable. Could you explain the fix and the bug you were talking about? Is it #815? What part of the old code was wrong? Would it be possible, that you temporary revert the code, add a test that fails due to the bug and then apply the fix again to see if this fixes the code? I could do or help you with that but I'd like to see a test for that quite subtile bug (if there was any). I also have 2 performance improvements added

File: src/buildings/nobMilitary.cpp [hunk 2]
```diff
@@ -329,60 +329,47 @@ void nobMilitary::LookForEnemyBuildings(const nobBaseMilitary* const exception)
...
-                // Wenns ein richtiges Militärgebäude ist, dann dort auch entsprechend setzen
-                if(BuildingProperties::IsMilitary((*it)->GetBuildingType()))
-                    static_cast<nobMilitary*>(*it)->NewEnemyMilitaryBuilding(1);
-            }
-            // andere Richtung muss auch getestet werden, zumindest wenns eine normaler Militärgebäude ist, Bug 389843
-            else if((*it)->GetGOT() == GOT_NOB_MILITARY)
+                newFrontierDistance = DIST_MID;
+            } else if((*it)->GetGOT() == GOT_NOB_MILITARY)
             {
                 nobMilitary* mil = dynamic_cast<nobMilitary*>(*it);
```
Comment (Flamefire) [inline src/buildings/nobMilitary.cpp]: Use safe_cast here to improve further (dynamic_cast in debug, static_cast in release mode)

File: src/buildings/nobMilitary.cpp [hunk 3]
```diff
@@ -329,60 +329,47 @@ void nobMilitary::LookForEnemyBuildings(const nobBaseMilitary* const exception)
...
-    }
 
-    // Evtl. Hafenpunkte in der N? mit ber?htigen
-    if(frontier_distance <= DIST_MID)
-        if(gwg->CalcDistanceToNearestHarbor(pos) < SEAATTACK_DISTANCE + 2)
-        {
-            // if(gwg->IsAHarborInSeaAttackDistance(MapPoint(x,y)))
-            frontier_distance = DIST_HARBOR;
+            // if new frontier distance is in military range, check if its reachable.
+            if(gwg->GetGGS().isEnabled(AddonId::FRONTIER_DISTANCE_REACHABLE) && newFrontierDistance >= DIST_MID
```
Comment (Flamefire) [inline src/buildings/nobMilitary.cpp]: Move that check (enabled) outside the loop to a `const bool`. The check is kinda costly (cache wise) due to a couple pointer indirections.

**Comment by Shawn8901** [comment]
Yes its #815 . i forgot to mention it in the pr, i only added a reference in the commit message. 
About the bug itself. 
the code before i had touched anything with the addon checked if the frontier distance is higher (within the loop). 
following could happen. if you have 3 buildings in buildings-array:
check building 1, its directly at the border --> seen as near.
check building 2, its in attackable range --> MID < NEAR, do nothing.
check building 3, building 3 is not able to reach the target (due unreachablility or to far away) since its the first check it will override frontier_distance with FAR and it will result in a FAR away military building. 

so the "bugfix" is not to override NEAR/MID with FAR. 
Rest are only some performance improvements and refactoring for the code.
So on my side i forgot that, because it was checked by a "magic"
if (!frontier_distance) clause, which idea was not to  override anything above FAR.

i think i will try for a test case, but as you remember i am (still) having trouble in understanding the world generation itself (see our discussions in the pr for this addon)

File: src/buildings/nobMilitary.cpp (hunk 2 above)
Comment (Shawn8901) [reply src/buildings/nobMilitary.cpp]: as i am no c++ professional, i don't now safe_cast as cast operation. after googling around i found only a Microsoft command extension which is named safe_cast.
i could do a static_cast if you want.

Comment (Flamefire) [reply src/buildings/nobMilitary.cpp]: It's something I introduced. Maybe I should rename it to something more common (some C++ helps use `down_cast` etc). The idea is the same: Use static_cast/dynamic_cast in a slim wrapper to capture the intent.

My version is `checkedCast` (https://github.com/Return-To-The-Roots/s25client/blob/master/common/include/commonDefines.h#L137) but I should enhance it... However it is to be used the same as `static_cast`: `safeCast<nobMilitary*>(src)` or similar.

**Comment by Shawn8901** [comment]
Finally i found some time for a test case. 
i have tried to build a test-case, which will cause the error and is as simple as possible.
the unit test builds an island, which is divided by a river, from 1/3 in the north to end of the map. Also i have added another river, which goes from west to east and is connected to the first river.
The map setup itself is designed, to have all military buildings picked up by LookForMilitaryBuildings, but have a long path to the other military buildings. 
 
now the test case, adds some buildings for player 1 (right side of the map), one is on the north side of the river and the other one is on the south side.  
After that a building for p0 is added on the north side of the river like in the screenshot.

image

everybody will assume, that the buildings on the north side, should be NEAR.
For current master code, that will not be the case. p0 will be FAR, since the dark green building is evaluated after the light green and its distance will be overriden (its a bit to  far away), so there will be no soldier able to reach the dark green building).

**Comment by Flamefire** [comment]
Perfect! Could you add a very short summary of this testcase to the testcase itself? You could use some simple ascii art to show the 2 rivers and Player 1&2. The testcases are not very readable by itself unfortunately and I don't think one can change that. A small (ASCII) map can help here. If thats to difficult add an image to the repo, name it properly and reference it in the testcase. I did that for some pathfinding stuff.

**Comment by Flamefire** [comment]
:+1: Keep up the good work!

**Pull request merged**